### ISSUE TITLE
`hot : true` and browser refreshing

### ISSUE THREAD

**Issue by piaste** [SAFE-Stack/SAFE-BookStore#197]
With the current release, hot reloading didn't fully work for my setup (no automatic browser refreshing) until, after finding an [old issue on webpack-dev-server](https://github.com/webpack/webpack-dev-server/issues/206), I removed the `hot : true` option from `webpack.config.js`. Apparently it conflicted somehow with `webpack.HotModuleReplacementPlugin()`.

If hot reloading and browser refreshing **currently** works for you, could you please check if it **still** works after removing `hot : true`? If so, it may be better for compatibility to remove it.

**Comment by forki** [comment]
/cc @maximemangel

Am 03.10.2017 13:01 schrieb "piaste" <notifications@github.com>:

**Comment by MangelMaxime** [comment]
@piaste Do you have `webpack-dev-server` installed globally ?

Removing the `hot: true` do not work for me. Can you please, try with this version: https://github.com/SAFE-Stack/SAFE-BookStore/pull/198

**Comment by piaste** [comment]
@MangelMaxime 
> Do you have webpack-dev-server installed globally ?

No. I have an older version of webpack installed globally, but no webpack-dev-server.

> Removing the hot: true do not work for me. Can you please, try with this version: #198

With those changes it doesn't work for me.

**Comment by MangelMaxime** [comment]
Can you please try to remove the global webpack ?
We need to launch the `webpack-dev-server` with the `hot: true` or `-- hot` option set. Otherwise, it will not start the HMR. You probably have something on your system that we don't have... I tested it on 3 computers, and several projects without problem... Don't really know how to debug it for you here.

**Comment by MangelMaxime** [comment]
@piaste Do you see `[WDS] Hot Module Replacement enabled.` in your console ?

And also, do you have any error in the network panel ?

**Comment by piaste** [comment]
@MangelMaxime 

After uninstalling the global Webpack, the CLI parameter (from your PR) triggers refreshing; the original still won't refresh, however.

This is the output when it _doesn't_ work (has been the same for every configuration I tried):

```
Initial state: Object
client?cd17:77 [WDS] Hot Module Replacement enabled.
program.fs:72 New message: Object
program.fs:74 Updated state: Object
program.fs:72 New message: Object
program.fs:74 Updated state: Object
2client?cd17:80 [WDS] App updated. Recompiling...
client?cd17:212 [WDS] App hot update...
log.js:23 [HMR] Checking for updates on the server...
log.js:23 [HMR] Updated modules:
log.js:23 [HMR]  - ./pages/WishList.fs
log.js:23 [HMR]  - ./App.fs
log.js:23 [HMR] App is up to date.
```

Honestly, after I fixed it on my end I sort of expected the `hot = true` option _must_ have been redundant in the presence of HMR, and that removing it would only smooth things. Didn't think it would open a rabbit hole 🤷‍♂️ and it's more of a webpack-dev-server issue in any case.

**Comment by MangelMaxime** [comment]
`hot: true` tell webpack-dev-server to serve file via HMR.

And the HMR plugin, tell webpack to add the plugin in the client so he know how to handle it. But yes, it's an issue with webpack-dev-server. Could you please open an issue on their repo explaining your situation, action etc. Perhaps, someone will be able to help

**Comment by piaste** [comment]
I belatedly realised something - deleting `hot = true` made autorefresh work on my end but *it wasn't actually using HMR*, just plain old full-page refresh, and I mistook it for HMR 🤦‍♂️ 

With the fix in your pull request, and after removing the global Webpack, I get *actual* HMR, so I think merging it would be better.

Not quite related, but in the current repo `App.fs` places `withDebugger` before `withReact`, and as your documentation states that doesn't work properly with the Redux devtools - you can look at the state history but not rewind/replay the page. Should I submit a PR to change it to:

```
#if DEBUG
|> Program.withConsoleTrace
|> Program.withHMR
#endif
|> Program.withReact "elmish-app"
#if DEBUG
|> Program.withDebugger
#endif
|> Program.run
```
?

**Comment by MangelMaxime** [comment]
@piaste Can you please check that having `hot = true` without the global webpack work for you ?

I would prefer, keep `hot = true` instead of using the `--hot` cli to keep the configuration in one place.

And yes, you should send a PR to fix the debugger support.

**Comment by kerams** [comment]
I'm seeing the exact same behavior--with the latest version of this repo I can't get a page refresh until I remove `hot = true` (at which point, as you said, HMR does not work). I've only recently installed Node + npm and Yarn and am positive that I didn't install Webpack (or anything else) manually myself.

**Comment by MangelMaxime** [comment]
@kerams are you sure that you got HMR and live reload? 

Can you please copy paste your browser console after doing an HMR update ?

**Comment by kerams** [comment]
```fsharp
Program.mkProgram init update view
|> Program.toNavigable (parseHash pageParser) urlUpdate
#if DEBUG
|> Program.withConsoleTrace
|> Program.withDebugger
|> Program.withHMR
#endif
|> Program.withReact "elmish-app"
|> Program.run
```

```
[WDS] App updated. Recompiling...
client?cd17:212 [WDS] App hot update...
log.js:23 [HMR] Checking for updates on the server...
log.js:23 [HMR] Updated modules:
log.js:23 [HMR]  - ./pages/WishList.fs
log.js:23 [HMR]  - ./App.fs
log.js:23 [HMR] App is up to date.
```

No refresh with `hot = true`.

**Comment by MangelMaxime** [comment]
@kerams Can you please tell me if you get HMR working with [this version](https://github.com/MangelMaxime/fable-suave-scaffold/tree/hot_via_cli) (take the hot_via_cli branch) ?

**Comment by kerams** [comment]
Your change makes no difference. However, I did find something curious. After I reload the page I get HMR and live reload on a menu item when logged out on the home page. When I go to the login page and change a menu item again, I no longer get a live reload on that page or home (!) until I type something in a field or navigate the menu.

**Comment by MangelMaxime** [comment]
If you remove this lines what's the result ?

```fs
|> Program.withConsoleTrace
|> Program.withDebugger
```

**Comment by kerams** [comment]
Still the same. After navigating away from the Home page, HMR only triggers after I force a render.

**Comment by kerams** [comment]
What's weirder is that if after all that I actually log in, I end up in some strange logged in/out state, where clicking on Wishlist takes me to the Home page and makes me appear logged out. Clicking on Login then makes me logged in instantly.

**Comment by kerams** [comment]
What's weirder is that if after all that I actually log in, I end up in some strange logged in/out state, where clicking on Wishlist takes me to the Home page and makes me appear logged out. Clicking on Login then makes me logged in instantly. I wonder if @piaste can reproduce any of this.

**Comment by piaste** [comment]
@kerams I'll try it this evening. I did all my attempts directly on the login page itself so didn't notice anything about that.

**Comment by purkhusid** [comment]
I have the same problem. I noticed that the views were not working were being rendered using the lazyView2 method and tried removing it. This fixed the HMR problem but I guess HMR should also work for lazy views?

**Comment by MangelMaxime** [comment]
@purkhusid HMR being `lazyView`, `lazyView2`, `lazyView3` works too but because they are lazy their will not re-render directly. They are lazy so they only re-render if your model changed, trigger a change into the submodel you passed to the `lazyView` and you will see that's the change have been applied to your application. We can't really do anything for that.

**Comment by dgroh** [comment]
still an issue? I tried `new webpack.HotModuleReplacementPlugin()` and also `hot: true` but none worked for me...

I use `"webpack-dev-server": "^2.11.1"` with `"webpack": "^3.11.0"`

**Comment by MangelMaxime** [comment]
@dgroh Are you using any of `lazyView`, `lazyView2`, `lazyView3` in your application ?

Do you have a global version of webpack isntalled ?
If you use [this project](https://github.com/MangelMaxime/Fulma-demo) do you get HMR ?

Can you please show us your `Program.mkProgram init update view` part ?

Which version of Fable.Elmish, Fable.Elmish.HMR, etc. are you using ?

**Comment by samzhao2008** [comment]
remove `hot:true` works for me.

I use `"webpack": "^4.29.3" ` and `"webpack-dev-server": "^3.1.14"`
